### ISSUE TITLE
Storage backend for low-end LAN devices 

### ISSUE THREAD

**Issue by fawick** [restic/restic#456]
)I'd like to propose a storage backend that is aimed for backing up to a super low-end device with very little CPU power in the local-area network, where backing up via SFTP is cumbersome as the CPU on the repository machine (an entry-level end-consumer NAS) is the bottleneck, while the network and the HDD would support much higher rates. In my particular use case, rsync over SSH (or restic over SFTP) is running at 2 MB/s, while writing via rsyncd results in up to 20 MB/s.  As the usecase is creating backups for machines over the trusted Ethernet at home, encryption of the transport layer is not needed. And the blobs themselves would be encrypted anyway. 

I already ruled out using NFS and exploiting a file system repository, as NFS is way slower than e.g. rsyncd and rsync. I'd assume that with a dedicated daemon, similar speeds are feasible with restic.

The storage backend would consist of:

1. an implementation of backend.Backend as a client, running on a reasonably powered CPU (e.g. Core i5)
2. a server daemon aimed on running on a low-end device such a consumer-grade entry-level NAS, e.g. a ARM5 CPU with 500-600 MHz and about 52 MB RAM (plus some hundred MB swap). The daemon might or might not be written in Go. 

Client and server would communicate via TCP in a pretty much netcat-like fashion, i.e., sending the raw blobs directly following a tiny header that contains the path and the length of the blob being written or read, so that the server-side CPU load is minimal. For the same reason, no compression of the transport itself would be used.

**Comment by mappu** [comment]
Would https://github.com/restic/restic/pull/253 meet this need ?

**Comment by fawick** [comment]
What I had in mind for a protocol between server and client is much leaner than HTTP.

What I failed to name in the proposal above: My use case is backing up high-res DSLR photos and videos, being at least a few Megabytes in size each. So maximizing throughput for many tiny files by introducing concurrency to the transfer is not a hard requirement here. Given my CPU and RAM constraints on the server side I would choose to have the transfer being done serially, reducing the amount of size of buffers. 

HTTP Pros:
- It's an open standard that is robust and scales well.
- @bchapuis defined the REST API, thereby taking care of what I called a "tiny header" anove.
- It is well-supported in the Go standard library.
- Optionally securing it later for WAN-usage with TLS and Authentication is straightforward and well-understood (even HTTP/2 nowadays).
- As with my proposal, the server side can be implemented in any language (though Go would still be my first choice, given how easy cross-compiling is)

HTTP Contras:
- A HTTP daemon needs to keep much more state in memory.
- The usual HTTP listener from the languages standard libraries will allow multiple parallel (well, at least concurrent) transfers. Though that can be balanced well on client-side.

So in summary, I do agree that investigating and contributing to #253 instead of implementing a new protocol may be the better choice than coming up with a yet another protocol and maturing it to the necessary robustness and scalability.
